test/object-store/sync/flx_role_change.cpp

```diff
@@ -802,57 +813,403 @@ TEST_CASE("flx: role changes during bootstrap complete successfully", "[sync][fl
+        auto& app_session = harness->session().app_session();
+        auto default_rule = app_session.admin_api.get_default_rule(app_session.server_app_id);
+
+        enum ClientResetTestState {
+            not_ready,
+            start,
+            // Primary sync session states before client reset
+            bind_before_cr_session,
+            // Fresh realm download sync session states
+            cr_session_ident,
+            cr_session_downloading,
+            cr_session_downloaded,
+            cr_session_integrating,
+            cr_session_integrated,
+            // Primary sync session states after fresh realm download
+            bind_after_cr_session,
+            merged_after_cr_session,
+            ident_after_cr_session,
+        };
+
+        bool client_reset_error = false;
+        bool role_change_error = false;
+        ClientResetTestState update_role_state = ClientResetTestState::not_ready;
+        int client_reset_count = 0;
+        bool skip_role_change_check = false;
+        TestingStateMachine<ClientResetTestState> client_reset_state(ClientResetTestState::not_ready);
+
+        // Set the state where the role change will be triggered
+        auto setup_test_params = [&](ClientResetTestState change_state, bool skip_role_check = false) {
+            client_reset_state.transition_with([&](ClientResetTestState) {
+                client_reset_error = false;       // Reset the client reset error tracking
+                role_change_error = false;        // Reset the role change error tracking
+                client_reset_count = 0;           // Reset the client reset error count
+                update_role_state = change_state; // State where the role change should be sent
+                // If the role change check is skipped, the test will not look for the role change error
+                // Depending on when the role change error is received (e.g. session deactivating), it
+                // may not be successfully or reliably captured with the event hook.
+                skip_role_change_check = skip_role_check;
+                return ClientResetTestState::start; // Update to start to begin tracking state
+            });
+        };
+
+        auto setup_config_callbacks = [&](SyncTestFile& config) {
+            // Use the sync client event hook to check for the error received and for tracking
+            // download messages and bootstraps
+            config.sync_config->on_sync_client_event_hook = [&](std::weak_ptr<SyncSession> session_ptr,
+                                                                const SyncClientHookData& data) {
+                bool is_fresh_path;
+                if (auto session = session_ptr.lock()) {
+                    is_fresh_path = _impl::client_reset::is_fresh_path(session->path());
+                }
+                else {
+                    //  Session is not valid anymore... exit now
+                    return SyncClientHookAction::NoAction;
```

Comment:
Is this expected to happen? The hook function getting called after the SyncSession is destroyed seems like it'd be indicating a bug somewhere.

Reply:
It's not supposed to happen - I updated it to a `REQUIRE(session)`